modules/assim.sequential/R/sda.enkf_refactored.R

```diff
@@ -0,0 +1,397 @@
+  CI.X2       <- CI.X1            # it was taken care of
+  #q.bar        <- NULL #default process covariance matrix
+  ##### Creating matrices that describe the bounds of the state variables
+  ##### interval is remade everytime depending on the data at time t
+  ##### state.interval stays constant and converts new.analysis to be within the correct bounds
+  interval    <- NULL
+  state.interval <- cbind(as.numeric(lapply(settings$state.data.assimilation$state.variables,'[[','min_value')),
+                          as.numeric(lapply(settings$state.data.assimilation$state.variables,'[[','max_value')))
+  rownames(state.interval) <- var.names
+  # weight matrix
+  wt.mat <- matrix(NA, nrow = nens, ncol = nt)
+  #Generate parameter needs to be run before this to generate the samples. This is hopefully done in the main workflow.
+  load(file.path(settings$outdir, "samples.Rdata"))  ## loads ensemble.samples
+  #reformatting params
+  params <- list()
+  for (i in seq_len(nens)) {
+    params[[i]] <- lapply(ensemble.samples, function(x, n) {
+      x[i, ] }, n = i)
+  } 
+  new.params <- params
+
+  ###-------------------------------------------------------------------###
+  ### If this is a restart - Picking up were we left last time          ###
+  ###-------------------------------------------------------------------###----   
+  if (restart){
+    load(file.path(settings$outdir,"SDA", "sda.output.Rdata"))
+    #--- Updating the nt and etc
+    if(!dir.exists(file.path(settings$outdir,"SDA",assimyears[t]))) dir.create(file.path(settings$outdir,"SDA",assimyears[t]))
+    # finding/moving files to it's end year dir
+    files.last.sda<-list.files.nodir(file.path(settings$outdir,"SDA"))
+    #copying
+       file.copy(file.path(file.path(settings$outdir,"SDA"),files.last.sda),
+                 file.path(file.path(settings$outdir,"SDA"),paste0(assimyears[t],"/",files.last.sda))
+                 )
+       params<-new.params
+  }else{
+    t<-0
+  }
+
+  ###------------------------------------------------------------------------------------------------###
+  ### loop over time                                                                                 ###
+  ###------------------------------------------------------------------------------------------------###---- 
+  while(t<nt){
+    t<-t+1
+    # do we have obs for this time - what year is it ?
+    obs <- which(!is.na(obs.mean[[t]]))
+    obs.year <- year(names(obs.mean)[t])
+    ###-------------------------------------------------------------------------###
+    ###  Taking care of Forecast. Splitting /  Writting / running / reading back###
+    ###-------------------------------------------------------------------------###-----  
+    #- Check to see if this is the first run or not and what inputs needs to be sent to write.ensemble configs
+    if (t>1){
+      #removing old simulations
+      unlink(list.files(outdir,"*.nc",recursive = T,full.names = T))
```

Comment:
This is not the most efficient way possible for forcing the run.model converts it's out to *.nc specially in large ensembles which takes a lot of time to clean. I didn't know of any option forcing `start.model.runs` to behave how we wanted it . Other ideas are appreciated.

Reply:
Why wouldn't `start.model.runs` create new nc files? while SDA progresses iteratively, each new year's output should be converted to nc, what was happening in your case?

Reply:
Well I don't really know why, but after converting the outputs of the first run, it was kinda skipping the conversion and always reading those first nc files.

Reply:
have you checked whether you still experience this without this line of command? I mean, you were skipping split input before so sipnet was running all years at every iteration, so it wouldn't recreate new nc files. but now that's fixed, maybe this is not an issue anymore?

Reply:
Sure I can give it a try.

Reply:
This is weird. NO it doesn't convert the new nc files. I don't know why ?!! 
![capture](https://user-images.githubusercontent.com/7999941/44345197-9f3a1680-a460-11e8-839f-d6491fd5fbb1.PNG)

@araiho : any idea ? Did you say, you saw the same behavior from `PEcAn.remote::start.model.runs` too ?

Reply:
weird, this plot again looks like a forward run only? is split_inputs.SIPNET not working somehow?

Reply:
Yup, this is when I get without deleting nc files. 
So let's say I'm doing SDA for 1960-1965. In the first run I split the met to 1960-1965 so I run the model for all of those years and there is a nc file for each in the out dir. For the second round which I wanna do 1961 the model creates 1961.out but since the nc file is there form the first run (which covered the whole simulation period) 1961.nc does NOT get updated. 
I think this is what's happening here. But if I delete the nc files in each iteration it works all fine.

Reply:
then yes, this is an expected behaviour, model2netcdf wouldn't overwrite the existing nc unless `overwrite=TRUE`.. but is there a reason why you're running the whole period in the beginning? if you're doing SDA for 1960-1965, shouldn't the initial forecast be run up until 1960? 

if you have to run the whole 1960-1965 period from the beginning, then you can utilize the overwrite argument of model2netcdf instead of unlinking maybe

Reply:
That makes sense. 
I'll see how can I modify it using your comments before using it in the multi site.

Reply:
The original sda was getting the initial run from settings$run$model$start.date etc. So, one had to make sure they changed that so it wouldn't run the whole time period in the spin up. But, you could change that so it gets the spin up time period from the settings$sda instead? I can't remember if there was a specific reason we wanted to keep the spin up dates in the run section of the xml.